### PR TITLE
LPS-155631 Adapt FDS to only render with items when CurrentViewComponent is ready

### PR DESCRIPTION
**Pull request by rebsilva**
Related issue: https://issues.liferay.com/browse/LPS-155631

Hello guys, hope you are all doing well!

I'm from Bravo team and we are working with FDS to make a story as you can see [here](https://issues.liferay.com/browse/LPS-152180). The good news is that we managed to render the `FrontendDataSet` without using the explicit render method, looking like this:

```
<Card title={Liferay.Language.get('predefined-values')}>
		<div id="lfr-object-web__predefined-values-render-fds" >
				<FrontendDataSet {...props} />
		</div>
</Card>
```

But we faced some issues where the items inside the table wouldn't render, it actually breaks the whole page. After some investigation we realized that the problem was the `<CurrentViewComponent />` wasn't ready to be rendered since the `useEffect` that brings the table hasn't been called yet. So our proposal is to add a verification on the condition to only render the `<CurrentViewComponent />` when the component is ready.

If you have any concerns about this please let me know, thanks a lot!

fyi: @markocikos

### CODE REVIEW

File: modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/FrontendDataSet.js [hunk 1]
```diff
@@ -450,9 +450,10 @@ const FrontendDataSet = ({
 					value={selectedItemsValue.join(',')}
 				/>
 
-				{items?.length ||
-				overrideEmptyResultView ||
-				inlineAddingSettings ? (
+				{(items?.length ||
+					overrideEmptyResultView ||
+					inlineAddingSettings) &&
+				CurrentViewComponent ? (
```
Comment (markocikos) [inline modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/FrontendDataSet.js:456]: With this change, we are rendering `ClayEmptyState`, showing no results found, until the component is ready. This does not look right. Component should be immediately ready, unless we are loading from a remote URL, and then there `componentLoading`, correctly showing loading indicator until the loading is done.

I'll add this case to our sample module and try to replicate.

Comment (miltonmc) [reply modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/FrontendDataSet.js:456]: @marcoscv-work, on fist paint the `CurrentViewComponent` is not ready yet cause the the `view` into the `views` prop has no `component` yet, just the `contentRenderer` (`'table'` in our case).   The call to the `requestComponent()` is into a `useEffect`.

https://github.com/liferay-frontend/liferay-portal/blob/c62f96ddf0a32392d944273c5f11d5c2e49062ac/modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/FrontendDataSet.js#L341-L356

https://github.com/liferay-frontend/liferay-portal/blob/c62f96ddf0a32392d944273c5f11d5c2e49062ac/modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/FrontendDataSet.js#L202-L225

https://github.com/liferay-frontend/liferay-portal/blob/c62f96ddf0a32392d944273c5f11d5c2e49062ac/modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/views/index.js#L31-L39

We can set the component to be in a loading state until the request is done. Does it sounds good for you?

Comment (markocikos) [reply modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/FrontendDataSet.js:456]: > We can set the component to be in a loading state until the request is done. Does it sound good for you?

The code in `master`, that you quoted above, is already doing this. `componentLoading` is the state variable that determines the loading state. By the time it is `false` [here](https://github.com/liferay-frontend/liferay-portal/blob/master/modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/FrontendDataSet.js#L442), `CurrentViewComponent` should always be set. The check in this PR is unnecessary.

If direct inclusion of `<FrontendDataSet>` is buggy, there is something else going on. As I said, I'm working on extending our sample module with this case.

**Comment by markocikos** [comment]
@rebsilva @miltonmc please see #2338 

I was not able to replicate the issue that items inside the table don't render. What do you have different in your case, compared to the sample?

**Comment by rebsilva** [comment]
Hi @markocikos, thanks for your answer!

In our case, the are not requesting the items for the table through `apiURL`, we already have the items and we are sending to render immediately (we don't send the `apiURL` prop at all).

During our investigation (also with the help of @rodrigopaulino) we realized that the `<CurrentViewComponent>` is trying to render being `undefined`, since it still needs to get the table view component from this `useEffect` as @miltonmc said above:
```

	 useEffect(() => { 
	 	setComponentLoading(true); 
	  
	 	requestComponent().then((component) => { 
	 		if (isMounted()) { 
	 			setComponentLoading(false); 
	 			dispatch(updateViewComponent(activeViewName, component)); 
	 		} 
	 	}); 
	 }, [ 
	 	activeViewName, 
	 	dispatch, 
	 	isMounted, 
	 	requestComponent, 
	 	setComponentLoading, 
	 ]); 
```

If we try to render using the `apiURL` it works because on the first attempt of rendering, since the items haven't been retrieved yet, it didn't try to render the view at all because `dataLoading`  is `true`:

```
const view = !dataLoading && !componentLoading ? (
{...}
       {items?.length ||
	overrideEmptyResultView ||
	inlineAddingSettings ? (
			<CurrentViewComponent
			     frontendDataSetContext={FrontendDataSetContext}
			     items={items}
			     itemsActions={itemsActions}
			     style={style}
			     {...currentViewProps}
			/>
	)
{...}

)
```

On the second rendering when the items are ready, the table component is also ready (CurrentViewComponent) and it renders with no issues. That's why the other occurrences of FDS in the portal are not breaking because they don't pass the items prop directly as we are doing now, they use the `apiURL` prop.

The way we send the items directly to the FDS is breaking because it attempts to render in the first try without having the time to retrieve the table view component. Since this type of usage is going to be usual from now on, likewise rendering the FDS in a React component form, we need to find a solution for this kind of situation. We suggested one in this PR, but we would be pleased to know if there are better ways of doing this, or if we could help in anything. Thanks again for your attention on this matter!

**Comment by markocikos** [comment]
Thanks for clarifications @rebsilva!

I see. So, you are not only using FDS directly as a React component, but you are also using it as a controlled component. This makes a huge difference! The component was never designed to be used that way. Only basic rendering will work, and no other features that modify data: filters, pagination, sorting, etc. I would expect a lot of bugs like this one.

Are you working on the same solution as @danvb112? This is the second PR in a short time with this use case, see https://github.com/liferay-frontend/liferay-portal/pull/2249 

@dsanz As I said in https://github.com/liferay-frontend/liferay-portal/pull/2249#issuecomment-1143287547, it is product question if we want to support this use case.

Currently, the component is meant to be used in two ways: 
1. You provide an `apiURL` (headless data fetching).
2. You provide a `dataProviderKey` and implement an `FDSDataProvider`.

This would be a third way, where we don't have any data fetching within FDS. Instead, we would expose data fetching callbacks, controlled by a parent component, and passed in as props.

**Comment by dsanz** [comment]
We're having a slack discussion about this. As Marko said, current impl is not ready to handle these cases because FDS is "tied" to the existence of a URL to fetch/query items.

To enable FDS being used as controlled component, we have to move forward with some plan to incorporate changes in the component depending on the features we need, backed up by proper sample additions with tests. For instance, first, add those changes to make the FDS to render a provided list of items. Then enable item inline edition, then filtering/search, item creation, and so on. This is to illustrate the size of the task at hand and to make it clear this needs some coordination at product level as well.

**Pull request closed without merging**